### PR TITLE
DROOLS-3652: DMN popover - Enter to apply

### PR DESCRIPTION
**Pull request by danielzhe**
This PR also prevents that the TAB and SHIFT + TAB to exit from the pop-up, as discussed in the scrum meeting.

This PR also fixes this error, that I've found during the debug:
error

### CODE REVIEW

**Comment by jomarko** [comment]
jenkins please retest this

**Comment by jomarko** [comment]
@danielzhe the test failures doesn't seem to be random, failed also before by retest request, could you please have a look?

**Comment by danielzhe** [comment]
@jomarko I don't think is related with this PR, but I rebased with the master anyway.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java [hunk 1]
```diff
@@ -166,17 +189,122 @@ public NameAndDataTypePopoverViewImpl(final Input nameEditor,
...
+
+    public QName getCurrentTypeRef() {
+        return currentTypeRef;
+    }
+
+    public String getCurrentName() {
+        return currentName;
+    }
+
+    void setKeyDownListeners() {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java]: @manstis I would like to ask you for your advice. I checked [CellEditorControlsViewImpl](https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/controls/container/CellEditorControlsViewImpl.java#L84) which do very similar logic. What do you think about registering new keydown handlers form this PR at that place?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java [hunk 2]
```diff
@@ -166,17 +189,122 @@ public NameAndDataTypePopoverViewImpl(final Input nameEditor,
...
+                keyEvent.stopPropagation();
+            } else if (isEscapeKeyPressed(keyEvent)) {
+                reset();
+                hide(false);
+            }
+        }
+    }
+
+    boolean isTabKeyPressed(final KeyboardEvent event) {
+        return Objects.equals(event.key, TAB_KEY);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java]: Just brainstorming if we really need to introduce constants for tab, enter ...
Can't we compare like: `Objects.equals(event.code, String.valueOf(KeyCodes.KEY_TAB))` ?

**Comment by jomarko** [comment]
### Manual review note 1
When name change is confirmed by **Enter**, popover is closed, however the keyboard shortcuts stop working. I think if user decided to confirm changes with **Enter**, he also expects to use keyboard to control editor afterwards.

**Comment by jomarko** [comment]
### Manual review note 2
If user change Name, then Data type, but then decides to press **Esc** the changes are applied once grid is refreshed. See the [video](https://drive.google.com/open?id=1FUjbGYfn2Px6d1ww00WGTdce8wu1TKdL)

**Comment by jomarko** [comment]
### Manual review note 3
When user press **Esc** when data type dropdown was expanded, the dropdown is closed, then when user press **Esc** to close the popover, no action is taken.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java (hunk 2 above)
Comment (danielzhe) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java]: `KeyboardEvent` doesn't have public `code`.

**Comment by danielzhe** [comment]
@jomarko Changes applied! I also found that when the popup using keyboard, the popup was not being focused, so user had to click on the popup. Now when the popup is opened, the first input is focused, so user can open, edit and close with enter or esc without have to use the mouse.

@manstis I had to do some changes in `BootstrapSelectDropDownMonitor` due the Issue 1 reported by @jomarko in this PR. What was happening is that when you closed the opened dropdown using ESC, the event `hidden.bs.select` was not being called. This issue was not introduced by this PR, but for this PR works good I had to fix it. I talked with @karreiro and this was the solution we proposed for this. Let me know if you think this will break something.

apply

**Comment by manstis** [comment]
@danielzhe Code style failure:-
```
[INFO] There are 2 errors reported by Checkstyle 8.10.1 with /home/manstis/workspaces/git/manstis/forks/kie-wb-common/kie-wb-common-dmn/kie-wb-common-dmn-client/target/checkstyle-rules.xml ruleset.
[ERROR] src/test/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverImplTest.java:[22,8] (imports) UnusedImports: Unused import - elemental2.dom.KeyboardEvent.
[ERROR] src/test/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverImplTest.java:[33,15] (imports) UnusedImports: Unused import - org.mockito.Mockito.doReturn.
```

**Comment by danielzhe** [comment]
@manstis Sorry. Fixed.

**Comment by danielzhe** [comment]
@manstis I don't think this issue you reported is related, because I'm having the same issue with master without my changes. I'll be on PTO for a few days, when I get back I can check this and check the changes you proposed. :) Thank you!

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java [hunk 3]
```diff
@@ -160,23 +211,135 @@ public NameAndDataTypePopoverViewImpl(final Input nameEditor,
...
+            if (isEnterKeyPressed(keyEvent)) {
+                hide(true);
+                keyEvent.preventDefault();
+                onClosedByKeyboard();
+            } else if (isEscapeKeyPressed(keyEvent)) {
+                reset();
+                hide(false);
+                onClosedByKeyboard();
+            } else if (isTabKeyPressed(keyEvent)) {
+                if (keyEvent.shiftKey) {
```
[the remark this answers is not in the document]
Comment (danielzhe) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java]: Yep, I don't see another clear and simple way.

**Comment by danielzhe** [comment]
@jomarko The only way to not use obsolete code is doing a big refactor and I think is not appropriate for this PR.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverImpl.java [hunk 4]
```diff
@@ -82,4 +83,9 @@ public void show(final Optional<String> editorTitle) {
     public void hide() {
         binding.ifPresent(b -> view.hide());
     }
+
+    @Override
+    public void setOnClosedByKeyboardCallback(final Consumer callback) {
```
[the remark this answers is not in the document]
Comment (danielzhe) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverImpl.java]: Why?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java [hunk 5]
```diff
@@ -160,23 +211,135 @@ public NameAndDataTypePopoverViewImpl(final Input nameEditor,
...
+
+    EventListener getKeyDownEventListener() {
+        return (e) -> keyDownEventListener(e);
+    }
+
+    Button getTypeSelectorButton() {
+        return (Button) getElement().querySelector(TYPE_SELECTOR_BUTTON_SELECTOR);
+    }
+
+    Button getManagerButton() {
```
[the remark this answers is not in the document]
Comment (danielzhe) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverViewImpl.java]: Thank you.

**Comment by danielzhe** [comment]
@jomarko Changes applied. I think we need @karreiro approve before merge. Or not, since you and @manstis already reviewed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jomarko) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/NameAndDataTypePopoverImpl.java]: To explicitly say, what the callback takes as an argument. In this form, you can call `callback.accepet(123)`, `callback.accepet("abc")`, `callback.accepet(true)` ...

**Comment by danielzhe** [comment]
@manstis Merge, please.

**Comment by manstis** [comment]
@danielzhe Please let me know when green :-)

**Comment by danielzhe** [comment]
I know I represent 0,001% of the people, but just to let this registered here: I hated this SonarCloud. I think the test coverage is good enough. I have 90% of the new code covered but now I have to find stupid things to cover just to have the magic 80% of conditions covered. I think the current coverage is good enough and the time searching more to cover is just wasted time. @manstis @karreiro

**Comment by manstis** [comment]
AFAIK Sonar is a guide not a blocker... let's merge this when the build (excluding Sonar) is green.

**Comment by danielzhe** [comment]
@manstis Sonar is happy now and it's all green, ready to merge!

**Comment by manstis** [comment]
@danielzhe Can you please change use of `i` in loops to a more understandable variable name? ;-)

**Pull request merged**